### PR TITLE
fix(manifest): solid deleted-version cleanup (R2 trash before DB delete)

### PR DESCRIPTION
**Pull request by riderx**
## Summary (AI generated)

- Manifest cleanup now **checks R2 existence → moves to `deleted-after-7-days/` (7-day lifecycle) → only then deletes the DB row**, so we never lose track of objects before trash.
- Missing R2 objects are treated as already gone and the DB row is deleted; shared delta files still referenced by other versions are kept.
- Hardened `on_version_update` queue: 300s HTTP timeout, 30 retries for version queue, no silent ACK on cleanup failure, cron sweeper for stuck soft-deleted versions.
- Added `scripts/reclaim_deleted_version_manifests.ts` (single path: reclaim all stuck deleted manifests with the same R2-then-DB rules).
- Load tests cover 5000-file manifests and partial trash/DB failures.

## Motivation (AI generated)

Production `manifest` was ~8GB / ~10M rows, with ~5.8M rows stuck on soft-deleted versions because cleanup deleted DB rows first (or failed mid-flight) and the queue archived work without recovery.

## Business Impact (AI generated)

Stops unbounded Postgres/replica growth from soft-deleted bundles, keeps OTA delta objects recoverable via lifecycle trash, and makes deleted-version cleanup durable under large manifests.

## Test Plan (AI generated)

- [x] `bunx vitest run tests/on-version-update-cleanup.unit.test.ts tests/queue-consumer-message-shape.unit.test.ts`
- [ ] Apply migration `sweep_deleted_version_manifests` in prod
- [ ] Deploy functions / workers with queue timeout change
- [ ] Run `bun scripts/reclaim_deleted_version_manifests.ts` against prod to drain backlog
- [ ] Confirm `manifest` rows on `deleted=true` versions drop, then `VACUUM` as needed

Generated with AI

Made with [Cursor](https://cursor.com)

<a href="https://cubic.dev/pr/Cap-go/capgo.app/pull/2724?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added a scheduled sweep to automatically clear soft-deleted app versions that still have leftover manifest data.
  * Added a recovery script to reclaim unreferenced manifest-backed objects from storage.
* **Bug Fixes**
  * Improved manifest cleanup to prevent removing shared files still referenced elsewhere.
  * Added safer storage presence checks to avoid incorrect database deletions when objects are missing.
* **Reliability**
  * Increased timeout and retry budgeting for version-related background queue processing.
* **Tests**
  * Expanded unit tests to verify cleanup ordering, rollback behavior, and shared-reference edge cases.

### CODE REVIEW

**Comment by riderx** [comment]
@coderabbitai review

File: supabase/functions/_backend/utils/s3.ts [hunk 1]
```diff
@@ -141,10 +141,61 @@ function shouldUseSizeRangeFallback(size: number, headError: unknown): boolean {
...
+  try {
+    const client = initS3(c)
+    const url = await client.getPresignedUrl('HEAD', fileId)
+    const response = await fetch(url, { method: 'HEAD' })
+    await response.body?.cancel()
+
+    if (response.status === 404)
+      return 'absent'
+    if (response.status === 200)
+      return 'present'
```
[the remark this answers is not in the document]
Comment (riderx) [reply supabase/functions/_backend/utils/s3.ts:162]: ✅ Fixed in `833eebe06`: `getObjectPresence` HEAD fetch now uses `AbortSignal.timeout(10_000)` and still fails closed on timeout/unknown.

File: supabase/migrations/20260722095157_sweep_deleted_version_manifests.sql [hunk 2]
```diff
@@ -0,0 +1,145 @@
+-- Sweep soft-deleted app_versions
+
+CREATE INDEX IF NOT EXISTS idx_app_versions_deleted_with_manifest
+  ON public.app_versions (id)
+  WHERE deleted = true AND manifest_count > 0;
+
+CREATE INDEX IF NOT EXISTS idx_app_versions_deleted_at_id
+  ON public.app_versions (deleted_at, id)
+  WHERE deleted = true;
```
[the remark this answers is not in the document]
Comment (riderx) [reply supabase/migrations/20260722095157_sweep_deleted_version_manifests.sql:9]: Acknowledged. Supabase migrations run in a transaction, so `CREATE INDEX CONCURRENTLY` isn't viable here. These are small partial indexes on `app_versions` (not `manifest`); transactional build is acceptable for this change.

File: supabase/functions/_backend/triggers/queue_consumer.ts [hunk 3]
```diff
@@ -799,8 +806,9 @@ async function processQueue(c: Context, db: ReturnType<typeof getPgClient>, queu
     }
   }
 
+  const retryBudget = getQueueMaxReads(queueName)
```
[the remark this answers is not in the document]
Comment (riderx) [reply supabase/functions/_backend/triggers/queue_consumer.ts:810]: ✅ Fixed in `f13662751`: Discord actionable failures + alert formatting now use `getQueueMaxReads(queueName)` (30 for version queue).

File: scripts/reclaim_deleted_version_manifests.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  if (key.startsWith(TRASH_PREFIX))
+    return
+
+  const exists = await objectExists(s3, bucket, key)
+  if (!exists)
+    return
+
+  await s3.send(new CopyObjectCommand({
+    Bucket: bucket,
+    CopySource: `${bucket}/${key}`,
```
[the remark this answers is not in the document]
Comment (riderx) [reply scripts/reclaim_deleted_version_manifests.ts]: ✅ Fixed in `f13662751`: reclaim `CopySource` path segments are `encodeURIComponent`'d before copy.
